src/ninja.cc

```diff
@@ -1575,6 +1668,72 @@ NORETURN void real_main(int argc, char** argv) {
     exit((ninja.*options.tool->func)(&options, argc, argv));
   }
 
+  // Determine whether to setup a Jobserver pool. This depends on
+  // --jobserver or --jobserver-mode=MODE being passed on the command-line,
+  // or NINJA_JOBSERVER=MODE being set in the environment.
+  //
+  // This must be ignored if a tool is being used, or no/infinite
+  // parallelism is being asked.
+  //
+  // At the moment, this overrides any MAKEFLAGS definition in
+  // the environment.
+  std::unique_ptr<Jobserver::Pool> jobserver_pool;
+
+  do {
+    if (options.tool)  // Do not setup pool when a tool is used.
+      break;
+
+    if (config.parallelism == 1 || config.parallelism == INT_MAX) {
+      // No-parallelism (-j1) or infinite parallelism (-j0) was specified.
+      break;
+    }
+
+    if (config.jobserver_mode == Jobserver::Config::kModeNone) {
+      // --jobserver was not used, and NINJA_JOBSERVER is not set.
+      break;
+    }
+
+    if (config.verbosity >= BuildConfig::VERBOSE)
+      status->Info("Creating jobserver pool for %d parallel jobs",
+                   config.parallelism);
+
+    std::string err;
+    jobserver_pool = Jobserver::Pool::Create(
+        static_cast<size_t>(config.parallelism), config.jobserver_mode, &err);
+    if (!jobserver_pool.get()) {
+      if (config.verbosity > BuildConfig::QUIET)
+        status->Warning("Jobserver pool creation failed: %s", err.c_str());
+      break;
+    }
+
+    std::string makeflags = jobserver_pool->GetEnvMakeFlagsValue();
+
+    //  Set or override the MAKEFLAGS environment variable in
+    // the current process. This ensures it is passed to sub-commands
+    // as well.
+#ifdef _WIN32
+    // TODO(digit): Verify that this works correctly on Win32.
+    // this code assumes that _putenv(), unlike Posix putenv()
+    // does create a copy of the input string, and that the
+    // resulting environment is passed to processes launched
+    // with CreateProcess (the documentation only mentions
+    // _spawn() and _exec()).
+    std::string env = "MAKEFLAGS=" + makeflags;
+    _putenv(env.c_str());
+#else   // !_WIN32
+    setenv("MAKEFLAGS", makeflags.c_str(), 1);
+#endif  // !_WIN32
+
+  } while (0);
```

Comment:
Hm ... maybe better to use a function than this goto in disguise.

Reply:
Good idea. Done this both for client and pool setup. Please take a look. I also created `SetEnvironmentVariable()` and `ClearEnvironmentVariable()` functions. Let me know if you want to move these to `util.h`.